### PR TITLE
hydra-unstable: add missing perl deps

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

Follow-up for #132739.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/perl-packages.nix [hunk 1]
```diff
@@ -2150,6 +2150,21 @@ let
     };
   };
 
+  CatalystPluginPrometheusTiny = buildPerlPackage {
+    pname = "Catalyst-Plugin-PrometheusTiny";
+    version = "0.005";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:2155]:
```suggestion
    version = "0.006";
```

File: pkgs/top-level/perl-packages.nix [hunk 2]
```diff
@@ -2150,6 +2150,21 @@ let
...
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/S/SY/SYSPETE/Catalyst-Plugin-PrometheusTiny-0.005.tar.gz";
+      sha256 = "a42ef09efdc3053899ae007c41220d3ed7207582cc86e491b4f534539c992c5a";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:2158]:
```suggestion
      url = "mirror://cpan/authors/id/S/SY/SYSPETE/Catalyst-Plugin-PrometheusTiny-0.006.tar.gz";
      sha256 = "1hd2nv21c2mq3yvhsys4a276bdwr5igs4b73m99v7ymzpabvjf9b";
```

File: pkgs/top-level/perl-packages.nix [hunk 3]
```diff
@@ -17433,6 +17476,37 @@ let
     propagatedBuildInputs = [ IPCSignal ];
   };
 
+  PrometheusTiny = buildPerlPackage {
+    pname = "Prometheus-Tiny";
+    version = "0.007";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:17481]:
```suggestion
    version = "0.008";
```

File: pkgs/top-level/perl-packages.nix [hunk 4]
```diff
@@ -17433,6 +17476,37 @@ let
...
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/R/RO/ROBN/Prometheus-Tiny-0.007.tar.gz";
+      sha256 = "0ef8b226a2025cdde4df80129dd319aa29e884e653c17dc96f4823d985c028ec";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:17484]:
```suggestion
      url = "mirror://cpan/authors/id/R/RO/ROBN/Prometheus-Tiny-0.008.tar.gz";
      sha256 = "17d7b69chkcq8fprww6m15glndyb0lms2l0xjbnbw04q7f8ncskk";
```

File: pkgs/top-level/perl-packages.nix [hunk 5]
```diff
@@ -17433,6 +17476,37 @@ let
...
+    meta = {
+      homepage = "https://github.com/robn/Prometheus-Tiny";
+      description = "A tiny Prometheus client";
+      license = with lib.licenses; [ artistic1 gpl1Plus ];
+    };
+  };
+
+  PrometheusTinyShared = buildPerlPackage {
+    pname = "Prometheus-Tiny-Shared";
+    version = "0.023";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:17496]:
```suggestion
    version = "0.024";
```

File: pkgs/top-level/perl-packages.nix [hunk 6]
```diff
@@ -17433,6 +17476,37 @@ let
...
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/R/RO/ROBN/Prometheus-Tiny-Shared-0.023.tar.gz";
+      sha256 = "7c2c72397be5d8e4839d1bf4033c1800f467f2509689673c6419df48794f2abe";
```
Comment (stigtsp) [inline pkgs/top-level/perl-packages.nix:17499]:
```suggestion
      url = "mirror://cpan/authors/id/R/RO/ROBN/Prometheus-Tiny-Shared-0.024.tar.gz";
      sha256 = "02w24r4amk8dqlavg6abxap48hzxrbda48f1pmrqypgx1cz59g4g";
```

**Review by stigtsp — CHANGES_REQUESTED** [review]
`CatalystPluginPrometheusTiny`, `PrometheusTiny` and `PrometheusTinyShared` have newer versions.

Other than that, looks good to me.

**Comment by happysalada** [comment]
I have a quick question when you have a moment.
Did you run this on a darwin machine ?
I'm curious how you found the missing depencenies.
Just asking so I can help with the next update.

**Comment by happysalada** [comment]
Here is an example of the log of the stuff that fails
```
unpacking sources
unpacking source archive /nix/store/0l5dbkf9yh7l2xqqfr8w4yr6yssk61v6-Hash-SharedMem-0.005.tar.gz
source root is Hash-SharedMem-0.005
setting SOURCE_DATE_EPOCH to timestamp 1502477361 of file Hash-SharedMem-0.005/SIGNATURE
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
patching ./t/taint.t...
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
ld: unknown option: -mmacosx-version-min=10.12
Warning: ExtUtils::CBuilder not installed or no compiler detected
Proceeding with configuration, but compilation may fail during Build

Checking prerequisites...
  recommends:
    *  Devel::CallChecker is not installed

ERRORS/WARNINGS FOUND IN PREREQUISITES.  You may wish to install the versions
of the modules indicated above before proceeding with this installation

Run 'Build installdeps' to install missing prerequisites.

Created MYMETA.yml and MYMETA.json
Creating new 'Build' script for 'Hash-SharedMem' version '0.005'
Building Hash-SharedMem
probe system failed: can't compile innocuous program at ./lib/Hash/features.probe line 45.
```

**Comment by Ma27** [comment]
>  Did you run this on a darwin machine ?

Nope. I basically ran the VM test and then I saw which dependencies were missing. I basically copied them over from https://github.com/NixOS/hydra/blob/master/flake.nix then.

**Comment by stigtsp** [comment]
> ld: unknown option: -mmacosx-version-min=10.12
> Warning: ExtUtils::CBuilder not installed or no compiler detected

this might be fixed by `perlPreHook = lib.optionalString stdenv.isDarwin "export LD=$CC";`

**Comment by Ma27** [comment]
@ofborg test hydra

**Comment by happysalada** [comment]
Result of `nixpkgs-review pr 133482` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>perl532Packages.CatalystPluginPrometheusTiny</li>
    <li>perl532Packages.HashSharedMem</li>
    <li>perl532Packages.PrometheusTinyShared</li>
    <li>perl534Packages.CatalystPluginPrometheusTiny</li>
    <li>perl534Packages.HashSharedMem</li>
    <li>perl534Packages.PrometheusTinyShared</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>perl532Packages.DataRandom</li>
    <li>perl532Packages.PrometheusTiny</li>
    <li>perl534Packages.DataRandom</li>
    <li>perl534Packages.PrometheusTiny</li>
  </ul>
</details>

**Comment by happysalada** [comment]
Testing the proposed fix on darwin.

**Comment by happysalada** [comment]
It seems to be working, there is still a failure after that, I'm not sure if it's not just a flaky test

Here is the failure
```
Building Hash-SharedMem
clang -I/nix/store/hvfg331bylcqhd07383053adswfbjjwi-perl-5.32.1/lib/perl5/5.32.1/darwin-thread-multi-2leve
env LD_RUN_PATH=/nix/store/hvfg331bylcqhd07383053adswfbjjwi-perl-5.32.1/lib/perl5/5.32.1/darwin-thread-mul
clang -I/nix/store/hvfg331bylcqhd07383053adswfbjjwi-perl-5.32.1/lib/perl5/5.32.1/darwin-thread-multi-2leve
env LD_RUN_PATH=/nix/store/hvfg331bylcqhd07383053adswfbjjwi-perl-5.32.1/lib/perl5/5.32.1/darwin-thread-mul
...skipping...
t/magic.t .......... ^Mt/magic.t .......... 1/678 ^M                              ^Mt/magic.t .......... o
t/many.t ........... ^Mt/many.t ........... 1/1803 ^Mt/many.t ........... 28/1803 ^Mt/many.t ........... 6
t/method.t ......... ^Mt/method.t ......... 1/375 ^M                              ^Mt/method.t ......... o
t/mode.t ........... ^Mt/mode.t ........... 1/2323 ^M                                ^Mt/mode.t ..........
t/octet.t .......... ^Mt/octet.t .......... 1/756 can't open shared hash /private/tmp/nix-build-perl5.32.1
# Looks like your test exited with 92 just after 704.
^M                              ^Mt/octet.t .......... Dubious, test returned 92 (wstat 23552, 0x5c00)
Failed 52/756 subtests
t/op.t ............. ^Mt/op.t ............. 1/388 ^Mt/op.t ............. 241/388 ^M
t/perm.t ........... ^Mt/perm.t ........... 1/91 ^M                            ^Mt/perm.t ........... ok
t/pod_cvg.t ........ ^M                          ^Mt/pod_cvg.t ........ skipped: Test::Pod::Coverage not a
t/pod_syn.t ........ ^M                          ^Mt/pod_syn.t ........ skipped: Test::Pod not available
t/rename.t ......... ^Mt/rename.t ......... 1/44 ^M                            ^Mt/rename.t ......... ok
t/snapshot.t ....... ^Mt/snapshot.t ....... 1/89 ^M                            ^Mt/snapshot.t ....... ok
t/taint.t .......... ^Mt/taint.t .......... 1/834 ^M                              ^Mt/taint.t .......... o
t/tally.t .......... ^Mt/tally.t .......... 1/149 ^M                              ^Mt/tally.t .......... o
t/threads.t ........ ^Mt/threads.t ........ 1/44 ^Mt/threads.t ........ 3/44 ^M
t/tidy.t ........... ^Mt/tidy.t ........... 1/12 ^M                            ^Mt/tidy.t ........... ok
t/tie.t ............ ^Mt/tie.t ............ 1/134 ^M                              ^Mt/tie.t ............ o
t/version_synch.t .. ^Mt/version_synch.t .. 1/4 ^M                          ^Mt/version_synch.t .. ok

Test Summary Report
-------------------
t/octet.t        (Wstat: 23552 Tests: 704 Failed: 0)
  Non-zero exit status: 92
  Parse errors: Bad plan.  You planned 756 tests but ran 704.
Files=31, Tests=39796, 157 wallclock secs ( 2.66 usr  0.45 sys + 144.61 cusr  3.45 csys = 151.17 CPU)
Result: FAIL
Failed 1/31 test programs. 0/39796 subtests failed.
```

At least it builds.
@Ma27 could you apply the fix the HashSharedMem suggested by @stigtsp

**Comment by happysalada** [comment]
I'm trying again to see if the failure is persistent.

**Comment by happysalada** [comment]
One linker error I've missed before
```
perl5.32.1-Hash-SharedMem> Undefined symbols for architecture x86_64:
perl5.32.1-Hash-SharedMem>   "_HLBNzorFAJMYbPEjiEKkMFBaKqZMkqq", referenced from:
perl5.32.1-Hash-SharedMem>       _main in conftest1.o
perl5.32.1-Hash-SharedMem> ld: symbol(s) not found for architecture x86_64
perl5.32.1-Hash-SharedMem> clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
```
Not sure it's related.

**Comment by happysalada** [comment]
That is the test failing
```
perl5.32.1-Hash-SharedMem> t/octet.t .......... 1/756 can't open shared hash /private/tmp/nix-build-perl5.32.1-Hash-SharedMem-0.005.drv-0/rA5cK5erwN/t3foo: Illegal byte sequence at t/octet.t line 695.
```

**Comment by happysalada** [comment]
I propose something like
```
      # perl5.32.1-Hash-SharedMem> t/octet.t .......... 1/756 can't open shared hash /private/tmp/nix-build-perl5.32.1-Hash-SharedMem-0.005.drv-0/rA5cK5erwN/t3foo: Illegal byte sequence at t/octet.t line 695.
      doCheck = !stdenv.isDarwin;
```

Here is the source for the failing test
https://metacpan.org/release/ZEFRAM/Hash-SharedMem-0.005/source/t/octet.t#L695
I don't see anything that jumps out. With that and the proposed fix, this builds.

File: pkgs/top-level/perl-packages.nix [hunk 7]
```diff
@@ -9416,6 +9445,20 @@ let
...
   };
 
+  HashSharedMem = buildPerlModule {
+    pname = "Hash-SharedMem";
+    version = "0.005";
+    src = fetchurl {
+      url = "mirror://cpan/authors/id/Z/ZE/ZEFRAM/Hash-SharedMem-0.005.tar.gz";
+      sha256 = "324776808602f7bdc44adaa937895365454029a926fa611f321c9bf6b940bb5e";
+    };
+    buildInputs = [ ScalarString ];
```
Comment (happysalada) [inline pkgs/top-level/perl-packages.nix:9455]:
```suggestion
    buildInputs = [ ScalarString ];
     # perl5.32.1-Hash-SharedMem> t/octet.t .......... 1/756 can't open shared hash /private/tmp/nix-build-perl5.32.1-Hash-SharedMem-0.005.drv-0/rA5cK5erwN/t3foo: Illegal byte sequence at t/octet.t line 695.
      doCheck = !stdenv.isDarwin;
```

**Comment by happysalada** [comment]
Next I have a failure on TestLWPUserAgent that seems to be persistent too
```
#   Failed test 'network_fallback on instance'
#   at t/06-network-fallback.t line 100.
#          got: '404'
#     expected: '302'
# Looks like you failed 7 tests of 29.
^M                                                 ^Mt/06-network-fallback.t .................. Dubious, t
Failed 7/29 subtests
t/07-mask-response.t ..................... ^Mt/07-mask-response.t ..................... 1/? ^M
t/08-isa-coderef.t ....................... ^Mt/08-isa-coderef.t ....................... 1/? ^M
t/09-dispatch-to-request-method.t ........ ^Mt/09-dispatch-to-request-method.t ........ 1/? ^M
t/10-request-args-network.t .............. ^Mt/10-request-args-network.t .............. 1/? ^M
t/11-request-args-internal.t ............. ^Mt/11-request-args-internal.t ............. 1/? ^M
t/12-chained-method-call.t ............... ^Mt/12-chained-method-call.t ............... 1/? ^M
t/50-examples-application_client_test.t .. ^M                                                ^Mt/50-exampl
t/51-call_psgi.t ......................... ^Mt/51-call_psgi.t ......................... 1/? ^M
t/zzz-check-breaks.t ..................... ^Mt/zzz-check-breaks.t ..................... 1/3 ^M

Test Summary Report
-------------------
t/06-network-fallback.t                (Wstat: 1792 Tests: 29 Failed: 7)
  Failed tests:  6, 15-16, 19, 22, 24, 28
  Non-zero exit status: 7
Files=16, Tests=252, 14 wallclock secs ( 0.19 usr  0.09 sys +  4.75 cusr  4.65 csys =  9.68 CPU)
Result: FAIL
Failed 1/16 test programs. 7/252 subtests failed.
make: *** [Makefile:911: test_dynamic] Error 255
```
I'm trying to disable the tests on darwin there and building again.

**Comment by happysalada** [comment]
Next I have a failure on TestLWPUserAgent that seems to be persistent too
```
#   Failed test 'network_fallback on instance'
#   at t/06-network-fallback.t line 100.
#          got: '404'
#     expected: '302'
# Looks like you failed 7 tests of 29.
^M                                                 ^Mt/06-network-fallback.t .................. Dubious, t
Failed 7/29 subtests
t/07-mask-response.t ..................... ^Mt/07-mask-response.t ..................... 1/? ^M
t/08-isa-coderef.t ....................... ^Mt/08-isa-coderef.t ....................... 1/? ^M
t/09-dispatch-to-request-method.t ........ ^Mt/09-dispatch-to-request-method.t ........ 1/? ^M
t/10-request-args-network.t .............. ^Mt/10-request-args-network.t .............. 1/? ^M
t/11-request-args-internal.t ............. ^Mt/11-request-args-internal.t ............. 1/? ^M
t/12-chained-method-call.t ............... ^Mt/12-chained-method-call.t ............... 1/? ^M
t/50-examples-application_client_test.t .. ^M                                                ^Mt/50-exampl
t/51-call_psgi.t ......................... ^Mt/51-call_psgi.t ......................... 1/? ^M
t/zzz-check-breaks.t ..................... ^Mt/zzz-check-breaks.t ..................... 1/3 ^M

Test Summary Report
-------------------
t/06-network-fallback.t                (Wstat: 1792 Tests: 29 Failed: 7)
  Failed tests:  6, 15-16, 19, 22, 24, 28
  Non-zero exit status: 7
Files=16, Tests=252, 14 wallclock secs ( 0.19 usr  0.09 sys +  4.75 cusr  4.65 csys =  9.68 CPU)
Result: FAIL
Failed 1/16 test programs. 7/252 subtests failed.
make: *** [Makefile:911: test_dynamic] Error 255
```
I'm trying to disable the tests on darwin there and building again.

**Comment by happysalada** [comment]
Next I have a failure on `nix log /nix/store/d647781w73bfyzzxwn8vwjvj7nc5p7x7-nix-perl-2.4pre20210802_47e96bb.drv`
not sure this is supposed to be fixed, perhaps this is why hydra unstable doesn't build on darwin.

```
build flags: SHELL=/nix/store/715y5lnw12vdcc3hh9sc9vbkdi6c83x6-bash-4.4-p23/bin/bash
  GEN    Makefile.config
  GEN    lib/Nix/Store.cc
  CXX    lib/Nix/Store.o
  LD     lib/Nix/Store.dylib
ld: file not found: /System/Library/Frameworks/Security.framework/Versions/A/Security for architecture x86
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [mk/lib.mk:117: lib/Nix/Store.dylib] Error 1
```

**Comment by Ma27** [comment]
I'm seriously wondering if that's worth the effort: when we disable the tests entirely to work around a linker issue, it seems to me as if the build may pass, but the package won't work properly.
Also, Hydra is supposed to be run on NixOS (IIRC it won't be easy to get it to run on other Linux distros or even Darwin).

**Comment by stigtsp** [comment]
One approach could be to remove the tests that fail on darwin with something like:

```
  preCheck = lib.optionalString stdenv.isDarwin ''
      rm t/06-network-fallback.t
  '';


```

**Pull request merged**